File: medication_timer.py
# medication_timer.py
import sys, os, json, argparse
from datetime import datetime, timedelta
try:
    from colorama import init, Fore, Style
    init()
    COLORS = True
except ImportError:
    COLORS = False
    Fore = Style = type('', (), {'RESET_ALL':'', 'GREEN':'', 'RED':'', 'YELLOW':'', 'CYAN':''})()

DATA_FILE = "medications.json"

class Medication:
    def __init__(self, name, dosage, interval_hours, last_taken=None, med_id=None):
        self.id = med_id or 0
        self.name = name
        self.dosage = dosage
        self.interval_hours = interval_hours
        self.last_taken = last_taken or datetime.now().isoformat()
        self.history = []  # list of ISO timestamps

    def is_due(self):
        last = datetime.fromisoformat(self.last_taken)
        next_dose = last + timedelta(hours=self.interval_hours)
        return datetime.now() >= next_dose

    def next_dose_time(self):
        last = datetime.fromisoformat(self.last_taken)
        return last + timedelta(hours=self.interval_hours)

    def take_dose(self):
        now = datetime.now()
        self.last_taken = now.isoformat()
        self.history.append(now.isoformat())

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "dosage": self.dosage,
            "interval_hours": self.interval_hours,
            "last_taken": self.last_taken,
            "history": self.history
        }

    @classmethod
    def from_dict(cls, data):
        m = cls(data["name"], data["dosage"], data["interval_hours"],
                data["last_taken"], data["id"])
        m.history = data.get("history", [])
        return m

class TimerApp:
    def __init__(self):
        self.medications = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.medications = [Medication.from_dict(m) for m in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([m.to_dict() for m in self.medications], f, indent=2)

    def add_medication(self, name, dosage, interval):
        med_id = max([m.id for m in self.medications] + [0]) + 1
        m = Medication(name, dosage, interval, med_id=med_id)
        self.medications.append(m)
        self.save()
        print(f"💊 Added medication: {name} {dosage} every {interval}h (ID: {med_id})")

    def list_medications(self):
        if not self.medications:
            print("No medications.")
            return
        print("\n📋 Medications:")
        for m in self.medications:
            due = "🔴 DUE" if m.is_due() else f"⏳ {m.next_dose_time().strftime('%H:%M')}"
            status_color = Fore.RED if m.is_due() else Fore.GREEN
            if COLORS:
                print(f"  [{m.id}] {m.name} {m.dosage} – {status_color}{due}{Style.RESET_ALL}")
            else:
                print(f"  [{m.id}] {m.name} {m.dosage} – {due}")

    def take_dose(self, med_id):
        for m in self.medications:
            if m.id == med_id:
                m.take_dose()
                self.save()
                print(f"✅ Took {m.name} {m.dosage} at {datetime.now().strftime('%H:%M')}")
                return
        print(f"❌ Medication with ID {med_id} not found.")

    def check_due(self):
        due_list = [m for m in self.medications if m.is_due()]
        if not due_list:
            print("✅ No medications due right now.")
            return
        print("\n⏰ Due Medications:")
        for m in due_list:
            print(f"  [{m.id}] {m.name} {m.dosage} – overdue by {int((datetime.now() - m.next_dose_time()).total_seconds() / 60)} min")

    def history(self):
        if not self.medications:
            print("No medications.")
            return
        print("\n📜 History (last 5 doses per medication):")
        for m in self.medications:
            recent = m.history[-5:] if m.history else []
            print(f"  {m.name}:")
            if recent:
                for ts in recent:
                    dt = datetime.fromisoformat(ts)
                    print(f"    - {dt.strftime('%Y-%m-%d %H:%M')}")
            else:
                print("    (no history)")

def main():
    parser = argparse.ArgumentParser(description="Medication Timer")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("name")
    add_parser.add_argument("dosage")
    add_parser.add_argument("interval", type=float)

    subparsers.add_parser("list")
    take_parser = subparsers.add_parser("take")
    take_parser.add_argument("id", type=int)
    subparsers.add_parser("check")
    subparsers.add_parser("history")

    args = parser.parse_args()
    app = TimerApp()

    if args.cmd == "add":
        app.add_medication(args.name, args.dosage, args.interval)
    elif args.cmd == "list":
        app.list_medications()
    elif args.cmd == "take":
        app.take_dose(args.id)
    elif args.cmd == "check":
        app.check_due()
    elif args.cmd == "history":
        app.history()

if __name__ == "__main__":
    main()

File: medication_timer.cpp
// medication_timer.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

struct Medication {
    int id;
    string name;
    string dosage;
    double intervalHours;
    time_t lastTaken;
    vector<time_t> history;

    bool isDue() const {
        time_t next = lastTaken + (time_t)(intervalHours * 3600);
        return time(nullptr) >= next;
    }

    time_t nextDose() const {
        return lastTaken + (time_t)(intervalHours * 3600);
    }

    void takeDose() {
        time_t now = time(nullptr);
        lastTaken = now;
        history.push_back(now);
    }
};

class TimerApp {
private:
    vector<Medication> medications;
    string dataFile = "medications.json";

    void load() {
        ifstream f(dataFile);
        if (!f.is_open()) return;
        json j;
        f >> j;
        for (auto& item : j) {
            Medication m;
            m.id = item["id"];
            m.name = item["name"];
            m.dosage = item["dosage"];
            m.intervalHours = item["interval_hours"];
            string lastStr = item["last_taken"];
            // parse ISO time (simple)
            struct tm tm = {};
            strptime(lastStr.c_str(), "%Y-%m-%dT%H:%M:%S", &tm);
            m.lastTaken = mktime(&tm);
            for (auto& h : item["history"]) {
                string hs = h;
                struct tm tmh = {};
                strptime(hs.c_str(), "%Y-%m-%dT%H:%M:%S", &tmh);
                m.history.push_back(mktime(&tmh));
            }
            medications.push_back(m);
        }
    }

    void save() {
        json j = json::array();
        for (auto& m : medications) {
            char buf[30];
            strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", localtime(&m.lastTaken));
            string lastStr(buf);
            vector<string> histStr;
            for (time_t t : m.history) {
                char hbuf[30];
                strftime(hbuf, sizeof(hbuf), "%Y-%m-%dT%H:%M:%S", localtime(&t));
                histStr.push_back(string(hbuf));
            }
            j.push_back({
                {"id", m.id},
                {"name", m.name},
                {"dosage", m.dosage},
                {"interval_hours", m.intervalHours},
                {"last_taken", lastStr},
                {"history", histStr}
            });
        }
        ofstream f(dataFile);
        f << setw(2) << j << endl;
    }

public:
    TimerApp() { load(); }

    void addMedication(const string& name, const string& dosage, double interval) {
        int maxId = 0;
        for (auto& m : medications) if (m.id > maxId) maxId = m.id;
        Medication m;
        m.id = maxId + 1;
        m.name = name;
        m.dosage = dosage;
        m.intervalHours = interval;
        m.lastTaken = time(nullptr);
        medications.push_back(m);
        save();
        cout << "💊 Added medication: " << name << " " << dosage << " every " << interval << "h (ID: " << m.id << ")\n";
    }

    void list() {
        if (medications.empty()) { cout << "No medications.\n"; return; }
        cout << "\n📋 Medications:\n";
        for (auto& m : medications) {
            string due;
            if (m.isDue()) due = "\033[31m🔴 DUE\033[0m";
            else {
                char buf[10];
                time_t next = m.nextDose();
                strftime(buf, sizeof(buf), "%H:%M", localtime(&next));
                due = string("\033[32m⏳ ") + buf + "\033[0m";
            }
            cout << "  [" << m.id << "] " << m.name << " " << m.dosage << " – " << due << "\n";
        }
    }

    void takeDose(int id) {
        for (auto& m : medications) {
            if (m.id == id) {
                m.takeDose();
                save();
                char buf[10];
                time_t now = time(nullptr);
                strftime(buf, sizeof(buf), "%H:%M", localtime(&now));
                cout << "✅ Took " << m.name << " " << m.dosage << " at " << buf << "\n";
                return;
            }
        }
        cout << "❌ Medication with ID " << id << " not found.\n";
    }

    void checkDue() {
        vector<Medication*> due;
        for (auto& m : medications) if (m.isDue()) due.push_back(&m);
        if (due.empty()) { cout << "✅ No medications due right now.\n"; return; }
        cout << "\n⏰ Due Medications:\n";
        for (auto* m : due) {
            long overdue = (time(nullptr) - m->nextDose()) / 60;
            cout << "  [" << m->id << "] " << m->name << " " << m->dosage << " – overdue by " << overdue << " min\n";
        }
    }

    void history() {
        if (medications.empty()) { cout << "No medications.\n"; return; }
        cout << "\n📜 History (last 5 doses per medication):\n";
        for (auto& m : medications) {
            cout << "  " << m.name << ":\n";
            vector<time_t> recent = m.history;
            if (recent.size() > 5) recent = vector<time_t>(recent.end()-5, recent.end());
            if (recent.empty()) {
                cout << "    (no history)\n";
            } else {
                for (time_t t : recent) {
                    char buf[20];
                    strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M", localtime(&t));
                    cout << "    - " << buf << "\n";
                }
            }
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: medication_timer <command> [options]\n";
        return 1;
    }
    TimerApp app;
    string cmd = argv[1];
    if (cmd == "add") {
        if (argc < 5) { cerr << "add <name> <dosage> <interval>\n"; return 1; }
        app.addMedication(argv[2], argv[3], stod(argv[4]));
    } else if (cmd == "list") {
        app.list();
    } else if (cmd == "take") {
        if (argc < 3) { cerr << "take <id>\n"; return 1; }
        app.takeDose(stoi(argv[2]));
    } else if (cmd == "check") {
        app.checkDue();
    } else if (cmd == "history") {
        app.history();
    } else {
        cerr << "Unknown command.\n";
    }
    return 0;
}
